### ISSUE TITLE
ILSpy shows the wrong scope for a hand-generated local variable

### ISSUE THREAD

**Issue by 0x53A** [icsharpcode/ILSpy#662]
This IL (it was generated by post processing with Cecil)

```
    .try
    {
        IL_0110: ldarg.0
        IL_0111: ldarg.0
        IL_0112: ldfld class [System.Data]System.Data.IDbCommand class Dapper.SqlMapper/'<QueryImpl>d__61`1'<!T>::'<cmd>5__4'
        IL_0117: dup
        IL_0118: isinst [System.Data]System.Data.IDbCommand
        IL_011d: stloc.s 11
        IL_011f: ldarg.0
        IL_0120: ldfld bool class Dapper.SqlMapper/'<QueryImpl>d__61`1'<!T>::'<wasClosed>5__6'
        IL_0125: brtrue.s IL_012b

        IL_0127: ldc.i4.s 16
        IL_0129: br.s IL_012d

        IL_012b: ldc.i4.s 48

        IL_012d: nop
        IL_012e: ldloc.s 11
        IL_0130: ldstr "IDbCommand"
        IL_0135: ldstr "ExecuteReader"
        IL_013a: call class [mscorlib]System.IDisposable '<>Generated.SqlLogger'::Enter(class [System.Data]System.Data.IDbCommand, string, string)
        IL_013f: stloc.s 10
        IL_0141: callvirt instance class [System.Data]System.Data.IDataReader [System.Data]System.Data.IDbCommand::ExecuteReader(valuetype [System.Data]System.Data.CommandBehavior)
        IL_0146: ldloc.s 10
        IL_0148: callvirt instance void [mscorlib]System.IDisposable::Dispose()
        IL_014d: ldnull
        IL_014e: stloc.s 10
        IL_0150: stfld class [System.Data]System.Data.IDataReader class Dapper.SqlMapper/'<QueryImpl>d__61`1'<!T>::'<reader>5__5'
        IL_0155: leave.s IL_0163
    } // end .try
    finally
    {
        IL_0157: ldloc.s 10
        IL_0159: brfalse.s IL_0162

        IL_015b: ldloc.s 10
        IL_015d: callvirt instance void [mscorlib]System.IDisposable::Dispose()

        IL_0162: endfinally
    } // end handler
```

gets translated to this C#

```
            try
            {
                IDbCommand expr_117 = this.<cmd>5__4;
                IDbCommand dbCommand = expr_117 as IDbCommand;
                CommandBehavior arg_141_1 = this.<wasClosed>5__6 ? (CommandBehavior.SequentialAccess | CommandBehavior.CloseConnection) : CommandBehavior.SequentialAccess;
                IDisposable disposable = SqlLogger.Enter(dbCommand, "IDbCommand", "ExecuteReader");
                IDataReader arg_150_1 = expr_117.ExecuteReader(arg_141_1);
                disposable.Dispose();
                disposable = null;
                this.<reader>5__5 = arg_150_1;
            }
            finally
            {
                IDisposable disposable;
                if (disposable != null)
                {
                    disposable.Dispose();
                }
            }
```

In the C# it looks like there are actually two different local variables both named disposable, but in the IL they refer to one local (stloc 10 / ldloc 10).

I would have expected something like    

```
        IDisposable disposable;
        try
        {
            ...
            disposable = SqlLogger.Enter(dbCommand, "IDbCommand", "ExecuteReader");
            ...
            disposable.Dispose();
            disposable = null;
            ...
        }
        finally
        {
            if (disposable != null)
            {
                disposable.Dispose();
            }
        }
```

**Comment by 0x53A** [comment]
Ok, I think I know why this IL gets decompiled incorrectly:

There is no valid C# which can generate this IL, C# requires that the variable is initialized before reading it in the finally, so it would have a  [ ldnull, stloc 10 ] at the position of the variable declaration.

**Comment by weltkante** [comment]
IL methods can (and usually do) have a flag to initialize all variables to default values, so explicitly initializing them in IL is not always necessary.

**Comment by 0x53A** [comment]
Yes, but AFAIK it is always done in C#, which is why ILSpy stumbles when it is NOT done.

**Comment by 0x53A** [comment]
I just skimmed over the source, so might be completely wrong, but I think the issue is, that local variables are initialized to `SpecialOpCode.Uninitialized` here: https://github.com/icsharpcode/ILSpy/blob/2726336b3a56d343457b453d9be76b793ea2ebc0/ICSharpCode.Decompiler/FlowAnalysis/SsaFormBuilder.cs#L243

If the method has initlocals set, they should probably be initialized to null (for classes) / default (for structs)

**Comment by weltkante** [comment]
You are right, this sounds like ILspy isn't handling the initlocals flag correctly.

While fixing that would probably generate code that compiles it may not be enough for the pattern matcher to recognize this as a using-statement. In that case one would have to debug the [pattern matcher](https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler/Ast/Transforms/PatternStatementTransform.cs#L182) for the using statements.

However seeing that the first check of the pattern matcher is for an assignment, it may just be enough to have initlocals to be properly implemented to get the pattern matcher back on track.

Unfortunately I don't know how one would implement it properly, I don't have experience with that stage of the decompilation, I've only been looking at the pattern matchers.

**Comment by dgrunwald** [comment]
It's highly annoying to try to reassembly an IL fragment where the types of local variables and parameters are missing, so I'm not going to try this... but I'm pretty sure that this got fixed when we merged the newdecompiler branch.

The ILAst now keeps track of the original `initlocals` store, and the C# translation should declare the variable as `= default(T)` where necessary.
Although this test case looks more like variable splitting incorrectly thought the stores in the `try` would not be visible in the `finally`. That's probably also fixed, as we completely reimplemented that part of the decompiler.
